File: framework/languages/java/e2e/RegistrationCodec/Client/src/main/java/systems/zlink/e2e/registrationcodec/client/Scenarios/RcA4DiLifecycleScenario.java
package systems.zlink.e2e.registrationcodec.client.Scenarios;

import java.util.List;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioAssert;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioContext;
import systems.zlink.e2e.registrationcodec.shared.Contracts;

public final class RcA4DiLifecycleScenario {
    private RcA4DiLifecycleScenario() {
    }

    public static void run(ScenarioContext context) {
        List<Contracts.DiLifecycleRes> replies = List.of(
            context.server().post("/registration/di-filter-order").submit(Contracts.DiLifecycleRes[].class).toCompletableFuture().join().body());
        ScenarioAssert.ensure(replies.stream().map(Contracts.DiLifecycleRes::scopedId).distinct().count() == 3,
            "RC-A4 scoped dependency was not recreated per request: " + replies);
        ScenarioAssert.ensure(replies.stream().map(Contracts.DiLifecycleRes::singletonId).distinct().count() == 1,
            "RC-A4 singleton dependency changed between requests: " + replies);
        ScenarioAssert.ensure(replies.get(2).disposedCount() == 3,
            "RC-A4 dispose count mismatch: " + replies);
        ScenarioAssert.waitForEvidenceValueSuffix(context.evidence(), "DI", "DiLifecycle", ":di-2");
        System.out.println("scenario RC-A4 passed");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/client/Scenarios/RcA4DiLifecycleScenario.kt
package systems.zlink.e2e.kotlin.registrationcodec.client.scenarios

import systems.zlink.e2e.kotlin.registrationcodec.DiLifecycleRes
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioAssert
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioHttpClient

class RcA4DiLifecycleScenario(
    private val server: ScenarioHttpClient,
    private val assert: ScenarioAssert,
) {
    fun run() {
        val diReplies = server.post("/registration/di-lifecycle", null, Array<DiLifecycleRes>::class.java).toList()
        assert.that(diReplies.map { it.scopedId }.distinct().size == 3, "RC-A4 scoped dependency was not recreated")
        assert.that(diReplies.map { it.singletonId }.distinct().size == 1, "RC-A4 singleton dependency changed")
        assert.that(diReplies[2].disposedCount == 3, "RC-A4 dispose count mismatch")
        assert.that(
            server.waitForEvidence("DI", "DiLifecycleReq", "${diReplies[2].scopedId}:${diReplies[2].singletonId}:di-2")
                .any { it.value.endsWith(":di-2") },
            "RC-A4 DI evidence missing",
        )
        println("scenario RC-A4 passed")
    }
}
